x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/grouping/Categorize.java

```diff
@@ -156,4 +176,175 @@ public Expression field() {
+                    context
+                );
+                field = null;
+                GroupingKey result = new GroupingKey(
+                    mode,
+                    new GroupingKeyThing(resultOffset, evaluator, categorizer),
+                    context.blockFactory()
+                );
+                categorizer = null;
+                evaluator = null;
+                return result;
+            } finally {
+                Releasables.close(field, evaluator, categorizer);
+            }
+
+        }
+
+        @Override
+        public ElementType intermediateElementType() {
+            return ElementType.INT;
+        }
+
+        @Override
+        public GroupingAggregator.Factory valuesAggregatorForGroupingsInTimeSeries(int timeBucketChannel) {
+            throw new UnsupportedOperationException("not supported in time series");
+        }
+    }
+
+    private static class GroupingKeyThing implements GroupingKey.Thing {
+        private final int resultOffset;
+        private final CategorizeEvaluator evaluator;
+        private final TokenListCategorizer.CloseableTokenListCategorizer categorizer;
+
+        private GroupingKeyThing(
+            int resultOffset,
+            CategorizeEvaluator evaluator,
+            TokenListCategorizer.CloseableTokenListCategorizer categorizer
+        ) {
+            this.resultOffset = resultOffset;
+            this.evaluator = evaluator;
+            this.categorizer = categorizer;
+        }
+
+        @Override
+        public Block evalRawInput(Page page) {
+            return evaluator.eval(page);
+        }
+
+        @Override
+        public int extraIntermediateBlocks() {
+            return 1;
+        }
+
+        @Override
+        public void fetchIntermediateState(BlockFactory blockFactory, Block[] blocks, int positionCount) {
+            blocks[resultOffset + 1] = buildIntermediateBlock(blockFactory, positionCount);
+        }
+
+        @Override
+        public Block evalIntermediateInput(BlockFactory blockFactory, Page page) {
+            BytesRefBlock intermediate = page.getBlock(resultOffset + 1);
+            Map<Integer, Integer> idMap;
+            if (intermediate.areAllValuesNull() == false) {
+                idMap = readIntermediate(intermediate.getBytesRef(0, new BytesRef()));
+            } else {
+                idMap = Collections.emptyMap();
+            }
+            IntBlock oldIds = page.getBlock(resultOffset);
+            try (IntBlock.Builder newIds = blockFactory.newIntBlockBuilder(oldIds.getTotalValueCount())) {
+                for (int i = 0; i < oldIds.getTotalValueCount(); i++) {
+                    newIds.appendInt(idMap.get(i));
+                }
+                return newIds.build();
+            }
+        }
+
+        @Override
+        public void replaceIntermediateKeys(BlockFactory blockFactory, Block[] blocks) {
+            // NOCOMMIT this offset can't be the same in the result array and intermediate array
```

Comment:
I bumped into this a few days ago and I think I need to dig some more - in this brave new world there's two "shapes" of data coming out of these grouping functions - the intermediate shape and the final shape. This is pretty similar to how aggs work - which is something I never fully understood to be honest. Anyway, I'm using `resultOffset` here - but that's the result offset of the intermediate data. not the *final* offset. So it can't be right.